File: java/leetcode_grind/Day337.java
package leetcode_grind;

import java.util.Arrays;

public class Day337 {
    // https://leetcode.com/problems/number-of-ways-to-stay-in-the-same-place-after-some-steps/description/
    static class Solution1 {
        public int numWays(int steps, int arrLen) {
            int MOD = 1000_000_007;
            arrLen = Math.min(arrLen, steps);
            var dp = new int[arrLen];
            var prevDp = new int[arrLen];
            prevDp[0] = 1;

            for (int remain = 1; remain <= steps; remain++) {
                dp = new int[arrLen];
                for (int curr = arrLen - 1; curr >= 0; curr--) {
                    var ans = prevDp[curr];
                    if (curr > 0) {
                        ans = (ans + prevDp[curr - 1]) % MOD;
                    }
                    if (curr < arrLen - 1) {
                        ans = (ans + prevDp[curr + 1]) % MOD;
                    }
                    dp[curr] = ans;
                }
                prevDp = dp;
            }

            return dp[0];
        }
    }

    static class Solution2 {
        public int paintWalls(int[] cost, int[] time) {
            var n = cost.length;
            var dp = new int[n + 1];
            var prevDp = new int[n + 1];
            Arrays.fill(prevDp, 1000000000);
            prevDp[0] = 0;

            for (int i = n - 1; i >= 0; i--) {
                dp = new int[n + 1];
                for (int remain = 1; remain <= n; remain++) {
                    var paint = cost[i] + prevDp[Math.max(0, remain - 1 - time[i])];
                    var dontPaint = prevDp[remain];
                    dp[remain] = Math.min(paint, dontPaint);
                }
                prevDp = dp;
            }
            return dp[n];
        }
    }
}

File: src/leetcode_grind/day_337.rs
pub fn num_ways(steps: i32, arr_len: i32) -> i32 {
    const MOD: i32 = 1000000007;
    let arr_len = arr_len.min(steps) as usize;
    let steps = steps as usize;
    let mut dp = vec![vec![0; steps + 1]; arr_len];
    dp[0][0] = 1;

    for remain in 1..=steps {
        for curr in (0..arr_len).rev() {
            dp[curr][remain] = (dp[curr][remain] + dp[curr][remain - 1]) % MOD;
            if curr > 0 {
                dp[curr][remain] = (dp[curr][remain] + dp[curr - 1][remain - 1]) % MOD;
            }
            if curr < arr_len - 1 {
                dp[curr][remain] = (dp[curr][remain] + dp[curr + 1][remain - 1]) % MOD;
            }
        }
    }

    dp[0][steps as usize]
}
